## 路径参数校验

可以使用 Path 为路径参数声明类型的校验和元数据。

## 导入 Path

首先，从 fastapi 导入 Path：

声明元数据，尽可能选择使用 Annotated 语法。

In [1]:
import uvicorn
from typing import Annotated
from fastapi import FastAPI, Path
app = FastAPI()
@app.get("/items/{item_id}")
async def read_items(
    item_id: Annotated[int, Path(title="The ID of the item to get")]
):
    results = {"item_id": item_id}
    return results

if __name__ == '__main__':
    config = uvicorn.Config(app, host='0.0.0.0', port=8009)
    server = uvicorn.Server(config)
    await server.serve()

INFO:     Started server process [18944]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8009 (Press CTRL+C to quit)


INFO:     127.0.0.1:49766 - "GET /items/5 HTTP/1.1" 200 OK


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [18944]


浏览器中输入 http://127.0.0.1:8009/items/5

会看到 {"item_id":5}

## 数值校验

添加 ge=5 后，item_id 将必须是一个大于（greater than）或等于（equal）5 的整数

In [10]:
import uvicorn
from typing import Annotated
from fastapi import FastAPI, Path
app = FastAPI()
@app.get("/items/{item_id}")
async def read_items(
    item_id: Annotated[int, Path(title="The ID of the item to get", ge=5)]
):
    results = {"item_id": item_id}
    return results

if __name__ == '__main__':
    config = uvicorn.Config(app, host='0.0.0.0', port=8009)
    server = uvicorn.Server(config)
    await server.serve()

INFO:     Started server process [18944]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8009 (Press CTRL+C to quit)


INFO:     127.0.0.1:56540 - "GET /items/4 HTTP/1.1" 422 Unprocessable Entity
INFO:     127.0.0.1:56539 - "GET /items/6 HTTP/1.1" 200 OK
INFO:     127.0.0.1:56544 - "GET /items/4 HTTP/1.1" 422 Unprocessable Entity


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [18944]


浏览器中输入 http://127.0.0.1:8009/items/6

会看到 {"item_id":6}

如果输入 http://127.0.0.1:8009/items/4
会看到 {"detail":[{"type":"greater_than_equal","loc":["path","item_id"],"msg":"Input should be greater than or equal to 5","input":"4","ctx":{"ge":5},"url":"https://errors.pydantic.dev/2.5/v/greater_than_equal"}]}

数值校验同样适用于 float 值。

In [11]:
import uvicorn
from typing import Annotated
from fastapi import FastAPI, Path
app = FastAPI()
@app.get("/items/{item_id}")
async def read_items(
    item_id: Annotated[float, Path(title="The ID of the item to get", ge=5.5)]
):
    results = {"item_id": item_id}
    return results

if __name__ == '__main__':
    config = uvicorn.Config(app, host='0.0.0.0', port=8009)
    server = uvicorn.Server(config)
    await server.serve()

INFO:     Started server process [18944]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8009 (Press CTRL+C to quit)


INFO:     127.0.0.1:56566 - "GET /items/5.6 HTTP/1.1" 200 OK
INFO:     127.0.0.1:56566 - "GET /items/5.4 HTTP/1.1" 422 Unprocessable Entity


INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [18944]


浏览器中输入 http://127.0.0.1:8009/items/5.6

会看到 {"item_id":5.6}

如果输入 http://127.0.0.1:8009/items/5.4
会看到 {"detail":[{"type":"greater_than_equal","loc":["path","item_id"],"msg":"Input should be greater than or equal to 5.5","input":"5.4","ctx":{"ge":5.5},"url":"https://errors.pydantic.dev/2.5/v/greater_than_equal"}]}

可选的声明数值校验：

gt：大于（greater tha
n）
ge：大于等于（greater than or equ
al）
lt：小于（less t
han）
le：小于等于（less than or equal）